skbio/sequence/_interval.py

```diff
@@ -0,0 +1,161 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2013--, scikit-bio development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+from __future__ import absolute_import, division, print_function
+from future.utils import with_metaclass
+
+import abc
+
+from ._feature import Feature
+from .intersection import Interval, IntervalTree
+import itertools
+from collections import defaultdict
+
+
+class IntervalMetadata():
+    def __init__(self, features=None):
+        # maps features attributes to intervals
+        if features is None:
+            self.feature_metadata = {}
+            self.interval_metadata = IntervalTree()
+        else:
+            self.interval_metadata = IntervalTree()
+            for k, invs in features.items():
+                for inv in invs:
+                    start, end = _polish_interval(inv)
+                    self.interval_metadata.add(start, end, k)
+            self.feature_metadata = features
+
+    def add(self, x, *intervals):
+        """ Adds a feature to the metadata object.
+
+        Parameters
+        ----------
+        x : skbio.sequence.feature
+            The feature object being added.
+        intervals : iteerable of intervals
+            A list of intervals associated with the feature
+
+        Note
+        ----
+        The intervals associated with a feature is assumed to be under
+        the `location` keyword by default.
+        """
+        if not isinstance(x, Feature):
+            raise ValueError('x is not an instance of `Feature`')
+
+        # TODO: The below will require careful consideration
+        # since this will be using a little more memory, since the whole
+        # feature object will be stored.
+        for interval in intervals:
+            loc = _polish_interval(interval)
+            if loc is not None:
+                start, end = loc
+                self.interval_metadata.add(start, end, x)
+
+        # TODO: The below will require careful consideration
+        # since this will be using a little more memory, since the whole
+        # feature object will be stored.
+        if x not in self.feature_metadata.keys():
+            self.feature_metadata[x] = []
+
+        self.feature_metadata[x] = list(map(_polish_interval, intervals))
+
+
+    def _query_interval(self, interval):
```

Comment:
right. and don't forget to add this case into tests when you fix it.

On Thu, Feb 11, 2016 at 10:33 PM, Jamie Morton <notifications@github.com>
wrote:

> In skbio/sequence/_interval.py
> <https://github.com/RNAer/scikit-bio/pull/1#discussion_r52709418>:
>
> > +        for interval in intervals:
> > +            loc = _polish_interval(interval)
> > +            if loc is not None:
> > +                start, end = loc
> > +                self.interval_metadata.add(start, end, x)
> > +
> > +        # TODO: The below will require careful consideration
> > +        # since this will be using a little more memory, since the whole
> > +        # feature object will be stored.
> > +        if x not in self.feature_metadata.keys():
> > +            self.feature_metadata[x] = []
> > +
> > +        self.feature_metadata[x] = list(map(_polish_interval, intervals))
> > +
> > +
> > +    def _query_interval(self, interval):
>
> [image: :+1:] Nice catch.
>
> The same thing will probably also happen in _query_features.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/RNAer/scikit-bio/pull/1/files#r52709418>.
>